pylot/planning/frenet_optimal_trajectory/fot_planning_operator.py

```diff
@@ -0,0 +1,259 @@
+        vy = can_bus_msg.data.velocity_vector.y
+
+        # get distance from car to spline and projection
+        line = geom.LineString(zip(wx, wy))
+        point = geom.Point(x, y)
+        distance = point.distance(line)
+        point_on_line = line.interpolate(line.project(point))
+
+        # compute tangent / normal spline vectors
+        x0, y0 = csp.calc_position(self.s0)
+        x1, y1 = csp.calc_position(self.s0 + 2)
+        svec = np.array([x1-x0, y1-y0])
+        svec = svec / np.linalg.norm(svec)  # unit vector tangent to spline
+        tvec = np.array([svec[1], -svec[0]])  # unit vector orthog. to spline
+
+        # compute tangent / normal car vectors
+        fvec = np.array([vx, vy])
+        fvec = fvec / np.linalg.norm(fvec)  # unit vector tangent to velocity
+        bvec = np.array([point_on_line.x - x, point_on_line.y - y])
+        bvec = bvec / np.linalg.norm(bvec)  # unit vector between car, spline
+
+        # get initial conditions in frenet frame
+        s0 = csp.find_s(x, y, self.s0)  # course position
+        c_speed = can_bus_msg.data.forward_speed  # speed [m/s]
+        c_d = np.sign(np.dot(tvec, bvec)) * distance  # lateral position [m]
+        c_d_d = c_speed * np.dot(tvec, fvec)  # lateral speed [m\s]
+        c_d_dd = 0.0  # lateral acceleration [m\s]
+        # TODO (@fangedward) add IMU for lat. acc. when 0.9.7 is fixed
+
+        self.s0 = s0
+        return s0, c_speed, c_d, c_d_d, c_d_dd
+
+    @staticmethod
+    def _build_obstacle_list(vehicle_transform, prediction_msg):
+        """
+        Construct an obstacle list of proximal objects given vehicle_transform.
+        """
+        obstacle_list = []
+        # look over all predictions
+        for prediction in prediction_msg.predictions:
+            # use all prediction times as potential obstacles
+            for transform in prediction.trajectory:
+                global_obstacle = vehicle_transform * transform
+                obstacle_origin = [global_obstacle.location.x,
+                                   global_obstacle.location.y]
+                dist_to_ego = np.linalg.norm([vehicle_transform.location.x -
+                                              obstacle_origin[0],
+                                              vehicle_transform.location.y -
+                                              obstacle_origin[1]])
+                # TODO (@fangedward): Fix this hack
+                # Prediction also sends a prediction for ego vehicle
+                # This will always be the closest to the ego vehicle
+                # Filter out until this is removed from prediction
+                if dist_to_ego < 2:  # this allows max vel to be 20m/s
+                    break
+                elif dist_to_ego < DEFAULT_DISTANCE_THRESHOLD:
+                    obstacle_list.append(obstacle_origin)
+        return np.array(obstacle_list)
+
+    def _update_global_waypoints(self, vehicle_transform):
```

Comment:
Please remove the method, and instead invoke compute_waypoints directly in the code. The method is too shallow :)

Reply:
Done